# Project to simulate a data set using numpy.random

#### The real world phenomenom I have chosen for this project is the gender breakdown at senior management level in Galway Mayo Institute of Technology.   
I will look at the following variables:   
Age   
Qualification   
Years of Experience   
Grade   
Gender      

The age variable will be a non-negative real number between 25 & 66.
The Qualification variable will be a real number with one of 3 different values: 8, 9, 10, to represent levels on the National Framework of Qualifications, (Level 8 = Honours Degree, Level 9 = Masters Degree, Level 10 = PhD)   .
The years of experience variable will be a non-negative real number, and will be based on age and qualification.
The Grade variable will be a categorical variable with one of 7 different values: Assistant Lecturer, Lecturer, Senior Lecturer 1, Senior Lecturer 2, Senior Lecturer 3, Registrar, President
The Gender variable will be either male or female
The determinant of whether a staff member has reached senior management level will be their Grade - those at SL2, SL3, Registrar & President will be considered at Senior Management level
Only staff categorised as Academic will be considered in this project.   


In [2]:
#Import pandas & numpy.
import pandas as pd
import numpy as np
# Set the seed so that the numbers can be reproduced.
np.random.seed(0)  

Pandas is a software library for data maniupulation & analysis.    
Numpy is a python library useful for large, multi-dimensional arrays & matrices, along with functions to operate on arrays. Numpy.random is part of the numpy library and is what I will use to generate random data to simulate my dataframe.   
np.random.seed sets the 'seed' for the random data. If the same seed is set the same numbers are generated each time. Changing the seed means a different set of numbers is generated.  

In [7]:
#Insert age, gender & qualification columns to dataframe
gender = ['male', 'female']
grade = ['Assistant Lecturer', 'Lecturer', 'Senior Lecturer 1', 'Senior Lecturer 2', 'Senior Lecturer 3', 'Registrar', 'President']
df = pd.DataFrame({'Age': np.random.randint(25,66,450),
                   'Gender': np.random.choice(gender, 450,p=[0.45, 0.55]), 
                   'Grade': np.random.choice(grade, 450, p=[.1573, .7584, .0225, .0393, .0169, .0028, .0028])})

#Function to assign qual depending on age, i.e. <27 cannot hold PhD
def qual(row):

                if row['Age']>= 27:

                    return np.random.choice (np.arange (8,11), p=[.62, .21, .17])
                
                else: 
                    
                    return np.random.choice (np.arange (8,10), p =[.666, .334])
                                  
qual =df.apply (lambda row: qual(row),axis=1)
df['Qualification'] = qual

#Function to determine experience based on age & qual
def exp(row):

                if row['Qualification']== 10:

                    return row['Age']-27
            
                elif row['Qualification']== 9:

                    return row['Age']-23
                
                elif row['Qualification']== 8: 
                    
                    return row['Age']-22
                                  
exp =df.apply (lambda row: exp(row),axis=1)
df['Experience'] = exp
df



Age  Gender               Grade  Qualification  Experience
0     64    male   Senior Lecturer 1             10          37
1     41    male            Lecturer              8          19
2     50    male            Lecturer             10          23
3     55  female            Lecturer              8          33
4     52  female            Lecturer              8          30
5     36  female            Lecturer             10           9
6     65  female            Lecturer             10          38
7     25  female            Lecturer              8           3
8     58  female            Lecturer             10          31
9     50    male   Senior Lecturer 2              8          28
10    39    male            Lecturer             10          12
11    59    male   Senior Lecturer 3             10          32
12    65  female            Lecturer             10          38
13    63    male            Lecturer              8          41
14    44  female            Lecturer              8          22
15    59  female  Assistant Lecturer              8          37
16    63    male            Lecturer              8          41
17    46    male            Lecturer              9          23
18    45  female            Lecturer              9          22
19    51    male            Lecturer              8          29
20    60    male            Lecturer              8          38
21    52    male            Lecturer              8          30
22    28  female   Senior Lecturer 3              8           6
23    50    male            Lecturer              8          28
24    32    male            Lecturer              8          10
25    29  female            Lecturer              8           7
26    29    male            Lecturer              8           7
27    34  female            Lecturer             10           7
28    29    male            Lecturer              8           7
29    58  female            Lecturer              8          36
..   ...     ...                 ...            ...         ...
420   48  female            Lecturer              8          26
421   49  female            Lecturer              8          27
422   50  female  Assistant Lecturer              9          27
423   30  female            Lecturer             10           3
424   47  female            Lecturer              8          25
425   41  female  Assistant Lecturer              8          19
426   56  female            Lecturer             10          29
427   25    male            Lecturer              9           2
428   28    male  Assistant Lecturer              8           6
429   58    male            Lecturer              9          35
430   56    male            Lecturer              8          34
431   31    male            Lecturer              8           9
432   25    male            Lecturer              8           3
433   33  female  Assistant Lecturer              8          11
434   49  female            Lecturer              8          27
435   31  female            Lecturer              9           8
436   36  female            Lecturer              8          14
437   29    male  Assistant Lecturer             10           2
438   62    male            Lecturer              9          39
439   32    male  Assistant Lecturer              8          10
440   50  female            Lecturer              8          28
441   30  female            Lecturer              9           7
442   62    male            Lecturer              8          40
443   29    male            Lecturer              8           7
444   37  female  Assistant Lecturer              8          15
445   41    male   Senior Lecturer 1              8          19
446   30  female            Lecturer              8           8
447   38    male            Lecturer              8          16
448   36    male            Lecturer              8          14
449   47  female            Lecturer              8          25

[450 rows x 5 columns]

### Variables
#### Age:
I used the function random.randint, a discrete uniform distribution, to generate ages of academic staff, with a minimum age of 25 and a maxiumum age of 66. It is appropriate to use this type of distribution when you have a minimum & maximum value, but no insight into where within this range a particular value might fall. I made the assumption, based on minimum requirements for academic posts, that the youngest one can be appointed is 25, and currently one must retire on 31st August after 65th birthday, making 66 the maximum age.  
#### Gender: 
I used np.random.choice to generate the data on gender, giving a 45%, 55% breakdown as per the July 2018 HEA Institutional Staff Profiles by Gender report. This function generates a random sample from a given one dimensional array, the parameters entered being gender, size of output, and probabilities associated with each entry of gender, i.e. 45% male, 55% female.   
#### Qualification:   
Initially I had used a random normal distribution to assign qualification, however having ran the code I realised that I need to look at the relationship between age & qualification, as, for example, a 25 year old staff member is unlikely to hold a PhD, given that the academic minimum requirements are a qualification to at least Level 8 plus three years experience. A Level 8 qualification generally takes 3/4 years, a Masters 1 year and a PhD at least 4 years. I therefore made the assumption that the youngest a PhD holder could be is 27. The IUA report on PhD students in the Universities found that the average age of a PhD student is 26 years (data collected in 2008), and a 1997 OECD report found that the typical graduation age from PhDs in Ireland is 24-27 years (Organization for Economic Co-operation and Development, Education at a Glance: OECD Indicators, 1997).   
I then had to create an if function to look at the age column, and if that was less than or equal to 27, randomly choose either Level 8 or Level 9 based on the current proportions of academic staff holding those qualifications - after researching the GMIT staff data I discovered that currently 21% of academic staff hold a PhD, 62% hold a Master's Degree, and the remainder an Honours Degree.  I applied the same function to those in the age column that were greater than 27, randomly choosing either a Level 8, 9 or 10 again using the current proportions of GMIT staff holding these qualifications.    

### References
https://nonpartisaneducation.org/Review/Resources/Int__lHigherEd_AppendixA.pdf   
http://www.iua.ie/wp-content/uploads/2013/08/IUS-01-2013-Report-on-PhD-Students.pdf   
http://hea.ie/assets/uploads/2018/01/Higher-Education-Institutional-Staff-Profiles-by-Gender-2018.pdf   
http://people.stern.nyu.edu/adamodar/New_Home_Page/StatFile/statdistns.htm   
https://docs.scipy.org/doc/numpy-1.15.1/reference/routines.random.html
